// @payunit/nodejs-sdk/dist/Collections.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collections = void 0;
const types_1 = require("./types");
const collect_1 = require("./validators/collect");
class Collections {
    constructor(client) {
        this.client = client;
    }
    initiatePayment(initiatePaymentRequest) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.client.makeRequest('POST', '/gateway/initialize', collect_1.initiatePaymentRequestSchema, initiatePaymentRequest, 'Failed to initiate payment');
        });
    }
    makePayment(makePaymentRequest) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateMobileMoneyProvider({
                provider: makePaymentRequest.gateway,
                phone_number: makePaymentRequest.phone_number,
            });
            return this.client.makeRequest('POST', '/gateway/makepayment', collect_1.makePaymentMobileMoneyRequestSchema, makePaymentRequest, 'Failed to make payment');
        });
    }
    // implement method that will combine initiatePayment() and makePayment() for mobile money payments
    initiateAndMakePaymentMobileMoney(request) {
        return __awaiter(this, void 0, void 0, function* () {
            // Validate the request
            const { error } = collect_1.initiateAndMakePaymentMobileMoneyRequestSchema.validate(request);
            if (error) {
                throw new Error(`Validation error: ${error.details[0].message}`);
            }
            // Validate mobile money provider
            this.validateMobileMoneyProvider({
                provider: request.gateway,
                phone_number: request.phone_number,
            });
            // First, initiate the payment
            const initiatePaymentRequest = {
                total_amount: request.total_amount,
                currency: request.currency,
                transaction_id: request.transaction_id,
                return_url: request.return_url,
                notify_url: request.notify_url,
                payment_country: request.payment_country,
                pay_with: request.gateway,
                redirect_on_failed: request.redirect_on_failed,
                custom_fields: request.custom_fields,
            };
            const initiationResponse = yield this.initiatePayment(initiatePaymentRequest);
            // Then, make the mobile money payment
            const makePaymentRequest = {
                transaction_id: initiationResponse.transaction_id,
                amount: request.total_amount,
                currency: request.currency,
                gateway: request.gateway,
                phone_number: request.phone_number,
                return_url: request.return_url,
                notify_url: request.notify_url,
            };
            return this.makePayment(makePaymentRequest);
        });
    }
    getTransactionStatus(transaction_id) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.client.makeRequest('GET', `/gateway/paymentstatus/${transaction_id}`, collect_1.transactionIdSchema, { transaction_id }, `Failed to get payment status for transaction id ${transaction_id}`);
        });
    }
    payWithCard(payWithCardRequest) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.client.makeRequest('POST', '/gateway/initialize', collect_1.initiatePaymentRequestSchema, payWithCardRequest, 'Failed to pay with card');
        });
    }
    validateMobileMoneyProvider(mobileMoneyDetails) {
        const provider = mobileMoneyDetails.provider;
        if (!Object.values(types_1.RegisteredProviders).includes(provider)) {
            throw new Error(`Invalid mobile money provider. Supported providers: ${Object.values(types_1.RegisteredProviders).join(', ')}`);
        }
        const phone = mobileMoneyDetails.phone_number;
        if (!types_1.phoneRegex[provider].test(phone)) {
            throw new Error(`Invalid phone number for ${provider}`);
        }
    }
}
exports.Collections = Collections;
