// @payunit/nodejs-sdk/dist/validators/collect.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initiateAndMakePaymentMobileMoneyRequestSchema = exports.transactionIdSchema = exports.makePaymentMobileMoneyRequestSchema = exports.initiatePaymentRequestSchema = void 0;
const joi_1 = __importDefault(require("joi"));
const types_1 = require("./../types");
const utils_1 = require("../utils");
exports.initiatePaymentRequestSchema = joi_1.default.object({
    total_amount: joi_1.default.number().min(0.01).required(),
    currency: joi_1.default.string()
        .valid(...Object.values(types_1.SupportedCurrencies))
        .required(),
    transaction_id: joi_1.default.string().required(),
    return_url: joi_1.default.string().pattern(utils_1.URL_PATTERN).required().optional(),
    notify_url: joi_1.default.string().pattern(utils_1.URL_PATTERN).required().optional(),
    payment_country: joi_1.default.string()
        .valid(...Object.values(types_1.PaymentCountry))
        .optional(),
    pay_with: joi_1.default.string()
        .valid(...Object.values(types_1.RegisteredProviders))
        .optional(),
    redirect_on_failed: joi_1.default.string().valid('yes', 'no').optional(),
    custom_fields: joi_1.default.object().optional(),
});
exports.makePaymentMobileMoneyRequestSchema = joi_1.default.object({
    amount: joi_1.default.number().min(0.01).required(),
    gateway: joi_1.default.string()
        .valid(...Object.values(types_1.RegisteredProviders))
        .required(),
    currency: joi_1.default.string()
        .valid(...Object.values(types_1.SupportedCurrencies))
        .required(),
    transaction_id: joi_1.default.string().required(),
    phone_number: joi_1.default.string()
        .pattern(/^\d{8,10}$/)
        .required(),
    return_url: joi_1.default.string().pattern(utils_1.URL_PATTERN).required().optional(),
    notify_url: joi_1.default.string().pattern(utils_1.URL_PATTERN).required().optional(),
    paymentType: joi_1.default.string().optional().default('button'),
});
exports.transactionIdSchema = joi_1.default.object({
    transaction_id: joi_1.default.string().required(),
});
exports.initiateAndMakePaymentMobileMoneyRequestSchema = joi_1.default.object({
    total_amount: joi_1.default.number().min(0.01).required(),
    currency: joi_1.default.string()
        .valid(...Object.values(types_1.SupportedCurrencies))
        .required(),
    transaction_id: joi_1.default.string().required(),
    gateway: joi_1.default.string()
        .valid(...Object.values(types_1.RegisteredProviders))
        .required(),
    phone_number: joi_1.default.string()
        .pattern(/^\d{8,10}$/)
        .required(),
    return_url: joi_1.default.string().pattern(utils_1.URL_PATTERN).optional(),
    notify_url: joi_1.default.string().pattern(utils_1.URL_PATTERN).optional(),
    payment_country: joi_1.default.string()
        .valid(...Object.values(types_1.PaymentCountry))
        .optional(),
    redirect_on_failed: joi_1.default.string().valid('yes', 'no').optional(),
    custom_fields: joi_1.default.object().optional(),
});
