// @payunit/nodejs-sdk/dist/validators/disbursement.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.disbursementPayTokenSchema = exports.disbursementConfirmRequestSchema = exports.disbursementRequestSchema = void 0;
const joi_1 = __importDefault(require("joi"));
const types_1 = require("../types");
exports.disbursementRequestSchema = joi_1.default.object({
    destination_currency: joi_1.default.string()
        .valid(...Object.values(types_1.SupportedCurrencies))
        .required(),
    debit_currency: joi_1.default.string()
        .valid(...Object.values(types_1.SupportedCurrencies))
        .required(),
    account_number: joi_1.default.string().required(),
    amount: joi_1.default.number().min(0).required(),
    beneficiary_name: joi_1.default.string().required(),
    deposit_type: joi_1.default.string().required(),
    transaction_id: joi_1.default.string().required(),
    country: joi_1.default.string()
        .valid(...Object.values(types_1.DisbursementPaymentCountry))
        .required(),
    account_bank: joi_1.default.string()
        .valid(...Object.values(types_1.RegisteredProviders))
        .required(),
});
exports.disbursementConfirmRequestSchema = joi_1.default.object({
    pay_token: joi_1.default.string().required(),
    deposit_message: joi_1.default.string().required(),
    deposit_note: joi_1.default.string().required(),
    notify_url: joi_1.default.string().optional(),
});
exports.disbursementPayTokenSchema = joi_1.default.object({
    pay_token: joi_1.default.string().required(),
});
