// @payunit/nodejs-sdk/dist/validators/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationError = void 0;
exports.validate = validate;
class ValidationError extends Error {
    constructor(details) {
        super(details.message);
        this.details = details;
        this.name = 'ValidationError';
    }
    static fromApiError(error) {
        if (error.status === 'FAILED' && error.statusCode === 400) {
            if (Array.isArray(error.message)) {
                return new ValidationError({
                    message: error.message.join(', '),
                    data: undefined,
                });
            }
            return new ValidationError({
                message: error.message,
                data: error.data,
            });
        }
        return new ValidationError({
            message: 'Unknow error occured with api',
            data: undefined,
        });
    }
    static fromJoiError(error) {
        return new ValidationError({
            message: error.details.map((detail) => detail.message).join(', '),
            data: error.details.map((detail) => detail.path.join('.')),
        });
    }
}
exports.ValidationError = ValidationError;
function validate(schema, data) {
    const { error } = schema.validate(data);
    if (error) {
        throw ValidationError.fromJoiError(error);
    }
    return data;
}
__exportStar(require("./collect"), exports);
__exportStar(require("./disbursement"), exports);
__exportStar(require("./invoice"), exports);
__exportStar(require("./checkout"), exports);
