// @payunit/nodejs-sdk/dist/types/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.configSchema = void 0;
const joi_1 = __importDefault(require("joi"));
const URL_PATTERN = /^https:\/\/[\w.-]+(?:\.[\w-]+)+[\w-._~:/?#[\]@!$&'()*+,;=]*$/;
exports.configSchema = joi_1.default.object({
    baseURL: joi_1.default.string().pattern(URL_PATTERN).default('https://gateway.payunit.net').messages({
        'string.pattern.base': 'baseURL must be a valid HTTPS URL',
    }),
    apiUsername: joi_1.default.string().required().min(3).max(100),
    apiPassword: joi_1.default.string().required().min(8).max(100),
    apiKey: joi_1.default.string().required().min(32).max(256),
    mode: joi_1.default.string().valid('test', 'live').default('test').required(),
    timeout: joi_1.default.number().integer().min(1000).max(60000).default(30000),
})
    .required()
    .options({ abortEarly: false });
