// @payunit/nodejs-sdk/dist/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedDepositType = exports.SupportedCurrencies = exports.TransactionType = exports.TransactionStatus = exports.phoneRegex = exports.DisbursementPaymentCountry = exports.PaymentCountry = exports.RegisteredProviders = void 0;
var RegisteredProviders;
(function (RegisteredProviders) {
    RegisteredProviders["CM_MTNMOMO"] = "CM_MTNMOMO";
    RegisteredProviders["CM_ORANGE"] = "CM_ORANGE";
    RegisteredProviders["WORLD_VISA"] = "WORLD_VISA";
    RegisteredProviders["GA_MOOV"] = "GA_MOOV";
    RegisteredProviders["GA_AIRTEL"] = "GA_AIRTEL";
    RegisteredProviders["CI_MTN"] = "CI_MTN";
    RegisteredProviders["CI_ORANGE"] = "CI_ORANGE";
    RegisteredProviders["CI_MOOV"] = "CI_MOOV";
    RegisteredProviders["SN_ORANGE"] = "SN_ORANGE";
    RegisteredProviders["SN_FREEMONEY"] = "SN_FREEMONEY";
    RegisteredProviders["SN_EMONEY"] = "SN_EMONEY";
    RegisteredProviders["SL_ORANGE"] = "SL_ORANGE";
    RegisteredProviders["SL_AFRICELL"] = "SL_AFRICELL";
    RegisteredProviders["KE_MPESA"] = "KE_MPESA";
    RegisteredProviders["BF_ORANGE"] = "BF_ORANGE";
    RegisteredProviders["BF_MOBICASH"] = "BF_MOBICASH";
    RegisteredProviders["TZ_AIRTEL"] = "TZ_AIRTEL";
    RegisteredProviders["TZ_TIGO"] = "TZ_TIGO";
    RegisteredProviders["TZ_HALOPESA"] = "TZ_HALOPESA";
    RegisteredProviders["TZ_VODAFONE"] = "TZ_VODAFONE";
    RegisteredProviders["UG_VODAFONE"] = "UG_VODAFONE";
    RegisteredProviders["UG_MTN"] = "UG_MTN";
    RegisteredProviders["UG_AIRTEL"] = "UG_AIRTEL";
    RegisteredProviders["GH_MTN"] = "GH_MTN";
    RegisteredProviders["GH_TIGO"] = "GH_TIGO";
    RegisteredProviders["GH_VODAFONE"] = "GH_VODAFONE";
    RegisteredProviders["ZM_MTN"] = "ZM_MTN";
    RegisteredProviders["ZM_AIRTEL"] = "ZM_AIRTEL";
    RegisteredProviders["ZM_ZAMTEL"] = "ZM_ZAMTEL";
    RegisteredProviders["RW_AIRTEL"] = "RW_AIRTEL";
    RegisteredProviders["RW_MTN"] = "RW_MTN";
})(RegisteredProviders || (exports.RegisteredProviders = RegisteredProviders = {}));
var PaymentCountry;
(function (PaymentCountry) {
    PaymentCountry["CM"] = "CM";
    PaymentCountry["GA"] = "GA";
    PaymentCountry["CI"] = "CI";
    PaymentCountry["SN"] = "SN";
    PaymentCountry["SL"] = "SL";
    PaymentCountry["KE"] = "KE";
    PaymentCountry["BF"] = "BF";
    PaymentCountry["TZ"] = "TZ";
    PaymentCountry["UG"] = "UG";
    PaymentCountry["GH"] = "GH";
    PaymentCountry["ZM"] = "ZM";
    PaymentCountry["RW"] = "RW";
})(PaymentCountry || (exports.PaymentCountry = PaymentCountry = {}));
var DisbursementPaymentCountry;
(function (DisbursementPaymentCountry) {
    DisbursementPaymentCountry["CM"] = "CM";
})(DisbursementPaymentCountry || (exports.DisbursementPaymentCountry = DisbursementPaymentCountry = {}));
exports.phoneRegex = {
    CM_MTNMOMO: /^\d{9}$/,
    CM_ORANGE: /^\d{9}$/,
    GA_MOOV: /^0\d{8}$/,
    GA_AIRTEL: /^0\d{8}$/,
    CI_MTN: /^\d{9,10}$/,
    CI_ORANGE: /^\d{9,10}$/,
    CI_MOOV: /^\d{9,10}$/,
    SN_ORANGE: /\d{9}$/,
    SN_FREEMONEY: /\d{9}$/,
    SN_EMONEY: /\d{9}$/,
    SL_ORANGE: /^(23|30|33|7[0-9])\d{6}$/,
    BF_ORANGE: /\d{8}$/,
    KE_MPESA: /\d{9}$/,
    BF_MOBICASH: /\d{8}$/,
    TZ_AIRTEL: /^\d{9}$/,
    TZ_TIGO: /^\d{9}$/,
    TZ_HALOPESA: /^\d{9}$/,
    TZ_VODAFONE: /^\d{9}$/,
    UG_VODAFONE: /^\d{9}$/,
    UG_MTN: /^\d{9}$/,
    UG_AIRTEL: /^\d{9}$/,
    GH_VODAFONE: /^[1-9]\d{8}$/,
    GH_MTN: /^[1-9]\d{8}$/,
    GH_TIGO: /^[1-9]\d{8}$/,
    RW_MTN: /^(78)\d{7}$/,
    RW_AIRTEL: /^\d{8}$/,
    ZM_AIRTEL: /^(9)\d{8}$/,
    ZM_MTN: /^(9)\d{8}$/,
    ZM_ZAMTEL: /^(9)\d{8}$/,
};
var TransactionStatus;
(function (TransactionStatus) {
    TransactionStatus["SUCCESS"] = "SUCCESS";
    TransactionStatus["SUCCESSFUL"] = "SUCCESSFUL";
    TransactionStatus["SUCCESSFULL"] = "SUCCESSFULL";
    TransactionStatus["FAILED"] = "FAILED";
    TransactionStatus["PENDING"] = "PENDING";
    TransactionStatus["CANCELLED"] = "CANCELLED";
    TransactionStatus["INITIATE"] = "INITIATE";
})(TransactionStatus || (exports.TransactionStatus = TransactionStatus = {}));
var TransactionType;
(function (TransactionType) {
    TransactionType["CARD"] = "CARD";
    TransactionType["MOBILE_MONEY"] = "MOBILE_MONEY";
})(TransactionType || (exports.TransactionType = TransactionType = {}));
var SupportedCurrencies;
(function (SupportedCurrencies) {
    SupportedCurrencies["USD"] = "USD";
    SupportedCurrencies["NGN"] = "NGN";
    SupportedCurrencies["EUR"] = "EUR";
    SupportedCurrencies["GBP"] = "GBP";
    SupportedCurrencies["XAF"] = "XAF";
    SupportedCurrencies["XOF"] = "XOF";
    SupportedCurrencies["EGP"] = "EGP";
    SupportedCurrencies["ZAR"] = "ZAR";
    SupportedCurrencies["MAD"] = "MAD";
    SupportedCurrencies["MWK"] = "MWK";
    SupportedCurrencies["MZN"] = "MZN";
    SupportedCurrencies["SLE"] = "SLE";
    SupportedCurrencies["KES"] = "KES";
    SupportedCurrencies["TZS"] = "TZS";
    SupportedCurrencies["UGX"] = "UGX";
    SupportedCurrencies["GHS"] = "GHS";
    SupportedCurrencies["ZMW"] = "ZMW";
    SupportedCurrencies["RWF"] = "RWF";
})(SupportedCurrencies || (exports.SupportedCurrencies = SupportedCurrencies = {}));
var SupportedDepositType;
(function (SupportedDepositType) {
    SupportedDepositType["MOBILE_MONEY"] = "MOBILE_MONEY";
    SupportedDepositType["BANK_TRANSFER"] = "BANK_TRANSFER";
})(SupportedDepositType || (exports.SupportedDepositType = SupportedDepositType = {}));
__exportStar(require("./config"), exports);
__exportStar(require("./collect"), exports);
__exportStar(require("./disbursement"), exports);
__exportStar(require("./invoice"), exports);
__exportStar(require("./checkout"), exports);
