// @payunit/nodejs-sdk/dist/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.URL_PATTERN = exports.Utils = void 0;
const crypto = __importStar(require("crypto"));
class Utils {
    constructor(encryptionKey, axiosInstance) {
        this.encryptionKey = encryptionKey;
        this.axiosInstance = axiosInstance;
    }
    // PCI DSS compliant card tokenization
    tokenizeCard(cardDetails) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                // Validate card data format
                this.validateCardDetails(cardDetails);
                // Tokenize card data
                const encryptedCardData = this.encryptPayload(cardDetails);
                const response = yield this.axiosInstance.post('/tokens/card', encryptedCardData);
                // Decrypt response
                const decryptedResponse = this.decryptResponse(response.data.encrypted, response.data.iv);
                return decryptedResponse;
            }
            catch (error) {
                if (error instanceof Error) {
                    throw new Error(`Failed to tokenize card: ${error.message}`);
                }
                throw new Error('Failed to tokenize card with unknown error');
            }
        });
    }
    // Validate card details according to PCI DSS requirements
    validateCardDetails(cardDetails) {
        if (!cardDetails.card_number ||
            !cardDetails.expiry_month ||
            !cardDetails.expiry_year ||
            !cardDetails.cvv) {
            throw new Error('Missing required card details');
        }
        // Basic validation
        if (!/^[0-9]{13,19}$/.test(cardDetails.card_number)) {
            throw new Error('Invalid card number format');
        }
        if (!/^[0-9]{2}$/.test(cardDetails.expiry_month) ||
            !/^[0-9]{4}$/.test(cardDetails.expiry_year)) {
            throw new Error('Invalid expiry date format');
        }
        if (!/^[0-9]{3,4}$/.test(cardDetails.cvv)) {
            throw new Error('Invalid CVV format');
        }
    }
    // Encrypt payload using AES-256
    encryptPayload(payload) {
        const iv = crypto.randomBytes(16);
        const cipher = crypto.createCipheriv('aes-256-cbc', Buffer.from(this.encryptionKey), iv);
        let encrypted = cipher.update(JSON.stringify(payload), 'utf8', 'hex');
        encrypted += cipher.final('hex');
        return { iv: iv.toString('hex'), encrypted };
    }
    // Decrypt response
    decryptResponse(encryptedData, iv) {
        const decipher = crypto.createDecipheriv('aes-256-cbc', Buffer.from(this.encryptionKey), Buffer.from(iv, 'hex'));
        let decrypted = decipher.update(encryptedData, 'hex', 'utf8');
        decrypted += decipher.final('utf8');
        return JSON.parse(decrypted);
    }
}
exports.Utils = Utils;
exports.URL_PATTERN = /^https:\/\/[\w.-]+(?:\.[\w-]+)+[\w-._~:/?#[\]@!$&'()*+,;=]*$/;
